napari/_qt/qt_resources/_svg.py

```diff
@@ -0,0 +1,149 @@
+"""
+from __future__ import annotations
+
+from functools import lru_cache
+
+from qtpy.QtCore import QByteArray, QPoint, QRect, QRectF, Qt
+from qtpy.QtGui import QIcon, QIconEngine, QImage, QPainter, QPixmap
+from qtpy.QtSvg import QSvgRenderer
+
+
+class QColoredSVGIcon(QIcon):
+    """A QIcon class that specializes in colorizing SVG files.
+
+    Parameters
+    ----------
+    path_or_xml : str
+        Raw SVG XML or a path to an existing svg file.  (Will raise error on
+        ``__init__`` if a non-existent file is provided.)
+    color : str, optional
+        A valid CSS color string, used to colorize the SVG. by default None.
+    opacity : float, optional
+        Fill opacity for the icon (0-1).  By default 1 (opaque).
+
+    Examples
+    --------
+    >>> from napari._qt.qt_resources import QColoredSVGIcon
+    >>> from qtpy.QtWidgets import QLabel
+
+    # Create icon with specific color
+    >>> label = QLabel()
+    >>> icon = QColoredSVGIcon.from_resources('new_points')
+    >>> label.setPixmap(icon.colored('#0934e2', opacity=0.7).pixmap(300, 300))
+    >>> label.show()
+
+    # Create colored icon using theme
+    >>> label = QLabel()
+    >>> icon = QColoredSVGIcon.from_resources('new_points')
+    >>> label.setPixmap(icon.colored(theme='light').pixmap(300, 300))
+    >>> label.show()
+    """
+
+    def __init__(
+        self, path_or_xml: str, color: str = None, opacity: float = 1.0
+    ) -> None:
+        from ...resources import get_colorized_svg
+
+        self._svg = path_or_xml
+        colorized = get_colorized_svg(path_or_xml, color, opacity)
+        super().__init__(SVGBufferIconEngine(colorized))
+
+    @lru_cache()
+    def colored(
+        self,
+        color: str = None,
+        opacity: float = 1.0,
+        theme: str = None,
+        theme_key: str = 'icon',
+    ):
+        """Return a new colorized QIcon instance.
+
+        Parameters
+        ----------
+        color : str, optional
+            A valid CSS color string, used to colorize the SVG.  If provided,
+            will take precedence over ``theme``, by default None.
+        opacity : float, optional
+            Fill opacity for the icon (0-1).  By default 1 (opaque).
+        theme : str, optional
+            Name of the theme to from which to get `theme_key` color.
+            ``color`` argument takes precedence.
+        theme_key : str, optional
+            If using a theme, key in the theme dict to use, by default 'icon'
+
+        Returns
+        -------
+        QColoredSVGIcon
+            A pre-colored QColoredSVGIcon (which may still be recolored)
+        """
+        if not color and theme:
+            from ...utils.theme import get_theme
+
+            color = get_theme(theme)[theme_key]
+
+        return QColoredSVGIcon(self._svg, color, opacity)
+
+    @classmethod
+    @lru_cache()
+    def from_resources(
+        cls,
+        icon_name: str,
+    ):
```

Comment:
Want to make it a static method for the lru cache to not take cls into account in the hashing ?

Reply:
ah interesting.  When will `cls` ever be different?  with subclasses?